dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/OpenTelemetryInfoFactoryImpl.java

```diff
@@ -0,0 +1,329 @@
+     *
+     * To ensure threads do not get the wrong OpenTelemetryInfo from the application metadata, it is stored as an AtomicReference
+     */
+
+    private static final String ENV_DISABLE_PROPERTY = "OTEL_SDK_DISABLED";
+    private static final String CONFIG_DISABLE_PROPERTY = "otel.sdk.disabled";
+    private static final String ENV_METRICS_EXPORTER_PROPERTY = "OTEL_METRICS_EXPORTER";
+    private static final String CONFIG_METRICS_EXPORTER_PROPERTY = "otel.metrics.exporter";
+    private static final String ENV_LOGS_EXPORTER_PROPERTY = "OTEL_LOGS_EXPORTER";
+    private static final String CONFIG_LOGS_EXPORTER_PROPERTY = "otel.logs.exporter";
+    private static final String SERVICE_NAME_PROPERTY = "otel.service.name";
+
+    private static final TraceComponent tc = Tr.register(OpenTelemetryInfoFactoryImpl.class);
+
+    private final MetaDataSlot slotForOpenTelemetryInfoHolder;
+
+    //This contains API calls that change between the upstream open telemetry version.
+    //We get a partially configued SDK Builder from OSGi becase we are in a static context
+    //and do not know which version of mpTelemetry will be in use.
+    @Reference
+    private OpenTelemetrySdkBuilderSupplier openTelemetrySdkBuilderSupplier;
+
+    @Activate
+    public OpenTelemetryInfoFactoryImpl(@Reference MetaDataSlotService slotService) {
+        slotForOpenTelemetryInfoHolder = slotService.reserveMetaDataSlot(ApplicationMetaData.class);
+    }
+
+    @Override
+    public OpenTelemetryInfo getOpenTelemetryInfo() {
+        try {
+            ApplicationMetaData metaData = ComponentMetaDataAccessorImpl.getComponentMetaDataAccessor().getComponentMetaData().getModuleMetaData().getApplicationMetaData();
+
+            return getOpenTelemetryInfo(metaData);
+        } catch (Exception e) {
+            Tr.error(tc, Tr.formatMessage(tc, "CWMOT5002.telemetry.error", e));
+            return new OpenTelemetryInfoImpl(false, OpenTelemetry.noop(), "unknown");
+        }
+    }
+
+    //A shortcut method to avoid fetching metadata more than we need to.
+    @Override
+    public OpenTelemetryInfo getOpenTelemetryInfo(ApplicationMetaData metaData) {
+        try {
+
+            //!!!!
+            //TODO Check if this is valid.
+            //Run io.openliberty.microprofile41.internal_fat to reproduce examples of liberty internal apps reaching this
+            //for "applications" that never went through application started.
+            //
+            String j2EEName = metaData.getJ2EEName().toString();
+            if (j2EEName.startsWith("io.openliberty") || j2EEName.startsWith("com.ibm.ws")) {
+                return new OpenTelemetryInfoImpl(false, OpenTelemetry.noop(), "Liberty internals");
+            }
+
+            OpenTelemetryInfoReference atomicRef = (OpenTelemetryInfoReference) metaData.getMetaData(slotForOpenTelemetryInfoHolder);
+            OpenTelemetryInfoWrappedSupplier supplier = atomicRef.get();
```

Comment:
As discussed, do the "is app internal check" after checking if `atomicRef` is null. Log a sensible exception if that occurs for a user app.

Reply:
Check